### PR TITLE
P0655R1 visit<R>()

### PR DESCRIPTION
**Pull request by NathanSWard**
# Description

This is an implementation of `std::visit<R>`.
Addresses issue #31 

# Checklist

- [x] I understand README.md. I also understand that acceptance of
  community PRs will be delayed until the test and CI systems are online.
- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before CI is online, leave this unchecked for
  initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
**I changed the underlying machinery of `std::variant` to not assume the return type to be `_Variant_visit_result_t` but rather have it be a templated parameter that is passed down by the calling `std::visit` function. I don't believe this causes ABI breaks, but am unsure.**
- [x] These changes were written from scratch using only this repository and
  the C++ Working Draft as a reference (and any other cited standards).
  If they were derived from a project that's already listed in NOTICE.txt,
  that's fine, but please mention it. If they were derived from any other
  project (including Boost and libc++, which are not yet listed in
  NOTICE.txt), you *must* mention it here, so we can determine whether the
  license is compatible and what else needs to be done.

### CODE REVIEW

**Comment by NathanSWard** [comment]
This is the example test case I used for this implementation. It directly mimics [P0355R1](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0655r1.pdf)'s example cases.
```
#include <variant>
#include <type_traits>

template <class I>
using O = I;

struct process {
    template<class I>
    constexpr auto operator()(I i) -> O<I> {
        return {};
    }
};

int main() {
    using I1 = int;
    using I2 = double;

    std::variant<I1, I2> input = {};

    auto output = std::visit<std::variant<O<I1>, O<I2>>>(process{}, input);
    auto result = std::visit<std::common_type_t<O<I1>, O<I2>>>(process{}, input);
    std::visit<void>(process{}, input);
}
```

File: stl/inc/variant [hunk 1]
```diff
@@ -39,6 +39,11 @@ struct _All_same : true_type {};
 template <class _First, class... _Rest>
 struct _All_same<_First, _Rest...> : bool_constant<conjunction_v<is_same<_First, _Rest>...>> {}; // variadic is_same
 
+template <class...>
+struct _All_covertible : true_type {};
```
Comment (sylveon) [inline stl/inc/variant]: Typo?

Comment (NathanSWard) [reply stl/inc/variant]: yes..., thanks!

**Comment by NathanSWard** [comment]
"fixup"-ed typo commit to original commit to make the entire commit easier to read.

File: stl/inc/variant [hunk 2]
```diff
@@ -39,6 +39,12 @@ struct _All_same : true_type {};
 template <class _First, class... _Rest>
 struct _All_same<_First, _Rest...> : bool_constant<conjunction_v<is_same<_First, _Rest>...>> {}; // variadic is_same
 
+template <class...>
+struct _All_convertible : true_type {};
+template <class _To, class... _From>
+struct _All_convertible<_To, _From...> : bool_constant<conjunction_v<is_convertible<_From, _To>...>> {
```
Comment (StephanTLavavej) [inline stl/inc/variant]: Naming: I am concerned that this reverses the `_From, _To` order. The reversal is necessary because of the variadic `_From`, but the name doesn't draw attention to that fact. (In metaprogramming, especially pre-concepts, we don't really have a type system to enforce proper usage, so conventions and naming are extremely important.)

Idea: `_Convertible_to_X_from_all`, or something else?

File: stl/inc/variant [hunk 3]
```diff
@@ -1601,6 +1626,51 @@ template <class _Ty>
...
+        _Size == 1 ? 0 : _Size <= 4 ? 1 : _Size <= 16 ? 2 : _Size <= 64 ? 3 : _Size <= 256 ? 4 : -1;
+    return _Visit_strategy<_Strategy>::template _Visit<_Ret, _ListOfIndexVectors>(
+        _Variant_visit_index1(0, static_cast<_As_variant<_Variants>&>(_Args)...), static_cast<_Callable&&>(_Obj),
+        static_cast<_As_variant<_Variants>&&>(_Args)...);
+}
+
+template <class _Callable, class... _Variants, class = void_t<_As_variant<_Variants>...>>
+constexpr _Variant_visit_result_t<_Callable, _As_variant<_Variants>...> visit(_Callable&& _Obj, _Variants&&... _Args) {
+    // Invoke _Obj with the contained values of _Args...
+    constexpr auto _Size = _Variant_total_states<_Remove_cvref_t<_As_variant<_Variants>>...>;
```
Comment (StephanTLavavej) [inline stl/inc/variant]: In `_HAS_CXX20` code, you can use C++20 `remove_cvref_t` instead of our `_Ugly` helper `_Remove_cvref_t` (needed for C++17 and earlier). Occurs below. (But perhaps @CaseyCarter would like the C++20 code to remain as similar as possible to the C++17 code.)

File: stl/inc/variant [hunk 4]
```diff
@@ -1601,6 +1626,51 @@ template <class _Ty>
...
+    using _ListOfIndexLists =
+        _Meta_list<_Meta_as_list<make_index_sequence<1 + variant_size_v<_Remove_cvref_t<_As_variant<_Variants>>>>>...>;
+    using _ListOfIndexVectors =
+        _Meta_transform<_Meta_quote<_Meta_as_integer_sequence>, _Meta_cartesian_product<_ListOfIndexLists>>;
+    static_assert(_Variant_all_visit_results_same<_Callable, _ListOfIndexVectors, _As_variant<_Variants>...>::value,
+        "visit() requires the result of all potential invocations to have the same type and value category "
+        "(N4828 [variant.visit]/2).");
```
Comment (StephanTLavavej) [inline stl/inc/variant]: The latest WP is N4835; please update your citations accordingly. Note that N4828 was mistakenly published and should not be cited; WG21-N4829 explains:

> Note: A working draft was circulated to the editing committee for review, and was mistakenly published with paper number N4828. N4828 is not the C++20 Committee Draft, and does not contain the results of addressing feedback from the editing committee.

File: stl/inc/variant [hunk 5]
```diff
@@ -1601,6 +1626,51 @@ template <class _Ty>
...
+
+    return _Visit_impl<_Size, _ListOfIndexVectors, _Variant_visit_result_t<_Callable, _As_variant<_Variants>...>>(
+        _STD forward<_Callable>(_Obj), _STD forward<_Variants>(_Args)...);
```
Comment (StephanTLavavej) [inline stl/inc/variant]: @CaseyCarter, does `<variant>` avoid `_STD forward<MEOW>(meow)` in favor of `static_cast<MEOW&&>(meow)` for throughput? (This is inconsistent with the rest of the STL, but we should at least be consistent within a single file.) I observe that there are no occurrences of `forward` in `<variant>` today.

File: stl/inc/variant [hunk 6]
```diff
@@ -1601,6 +1626,51 @@ template <class _Ty>
...
+}
+
+template <class _Ret, class _Callable, class... _Variants, class = void_t<_As_variant<_Variants>...>>
+constexpr _Ret visit(_Callable&& _Obj, _Variants&&... _Args) {
+    // Invoke _Obj with the contained values of _Args...
```
Comment (StephanTLavavej) [inline stl/inc/variant]: It seems like `visit<R>()` should have a different comment from `visit()`, mentioning `_Ret` somehow, or these comments should be removed.

File: stl/inc/variant [hunk 7]
```diff
@@ -1460,29 +1466,38 @@ struct _Variant_dispatcher;
 
 template <size_t... _Is>
 struct _Variant_dispatcher<index_sequence<_Is...>> {
-    template <class _Callable, class... _Types, bool _Any_valueless = ((_Is == 0) || ...)>
-    _NODISCARD static constexpr _Variant_visit_result_t<_Callable, _Types...> _Dispatch(
-        _Callable&& _Obj, _Types&&... _Args) {
+    template <class _Ret, class _Callable, class... _Types, bool _Any_valueless = ((_Is == 0) || ...)>
```
Comment (miscco) [inline stl/inc/variant]: Given that you change the world anyway would it make sense to use `std::disjunction` here as that short circuits? 

```cpp
bool _Any_valueless = disjunction_v<_Is == 0), ...>
```

Comment (NathanSWard) [reply stl/inc/variant]: If I'm not mistaken, I do believe fold expression already short circuit.

Comment (miscco) [reply stl/inc/variant]: As I understand it that is not the case and especially mentioned as a use case of {con, dis}junction:

https://en.cppreference.com/w/cpp/types/disjunction

Comment (miscco) [reply stl/inc/variant]: From the notes

> The short-circuit instantiation differentiates disjunction from fold expressions: a fold expression like (... || Bs::value) instantiates every B in Bs, while std::disjunction_v<Bs...> stops instantiation once the value can be determined.

Comment (NathanSWard) [reply stl/inc/variant]: Yes, when you use a fold expression such as `((_Is == 0) || ...)` every `_Is` is instantiated, however, when they they are being evaluated, short circuiting will still occur.
e.g.
```
template<std::size_t... Is>
constexpr bool b = ((Is == 0) || ...);

bool result = b<1, 0, 2>;
``` 
The fold expression will be instantiated such as
```
constexpr bool b = (1 == 0 || 0 == 0 || 2 == 0);
```
So each `Is` is instantiated, however, once `0 == 0` is evaluated as true, `2 == 0` will not be evaluated.
On the other hand, `std::disjunction` will only instantiate each value as needed.
**tldr:** Both will short circuit, it's just that the fold expression will instantiate every value.
Regardless, I do also prefer `std::disjunction` and will probably change it (thanks for the help :D).

File: stl/inc/variant (hunk 4 above)
Comment (NathanSWard) [reply stl/inc/variant]: Should I also update the C++17 `static_assert()` message or just the C++20 ones?

Comment (StephanTLavavej) [reply stl/inc/variant]: Updating those citations would help chip away at #182, go for it! 😸

File: stl/inc/variant (hunk 7 above)
Comment (CaseyCarter) [reply stl/inc/variant]: If the `_Is` were expensive to evaluate, then `disjunction` would make sense here. Given that the `_Is` are actually integer constants here, I doubt we can beat the compile time of this simple fold expression.

File: stl/inc/variant [hunk 8]
```diff
@@ -1460,29 +1466,38 @@ struct _Variant_dispatcher;
 
 template <size_t... _Is>
 struct _Variant_dispatcher<index_sequence<_Is...>> {
-    template <class _Callable, class... _Types, bool _Any_valueless = ((_Is == 0) || ...)>
-    _NODISCARD static constexpr _Variant_visit_result_t<_Callable, _Types...> _Dispatch(
-        _Callable&& _Obj, _Types&&... _Args) {
+    template <class _Ret, class _Callable, class... _Types, bool _Any_valueless = ((_Is == 0) || ...)>
+    _NODISCARD static constexpr _Ret _Dispatch(_Callable&& _Obj, _Types&&... _Args) {
```
Comment (CaseyCarter) [inline stl/inc/variant]: Our name-mangling doesn't track which types came from parameter packs - unlike the Itanium ABI (https://godbolt.org/z/1ubA_c) - so this could potentially link to an old/wrong specialization of `_Dispatch` in another TU. I can't imagine such a contrived situation happening in the real world, but we should change the name to `_Dispatch2` here nonetheless to avoid this miniscule ABI break.

File: stl/inc/variant [hunk 9]
```diff
@@ -1460,29 +1466,38 @@ struct _Variant_dispatcher;
...
             return _C_invoke(
                 static_cast<_Callable&&>(_Obj), _Variant_raw_get<_Is - 1>(static_cast<_Types&&>(_Args)._Storage())...);
+#endif // _HAS_CXX20
         }
     }
 };
 
-template <class _Ordinals, class _Callable, class _Variants>
+template <class _Ret, class _Ordinals, class _Callable, class _Variants>
 struct _Variant_dispatch_table; // undefined
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/variant]: This is fine - changing from 3 to 4 template type parameters results in distinct mangled names - but `_Variant_dispatcher::_Dispatch` is not. I'll audit.

File: stl/inc/variant [hunk 10]
```diff
@@ -1506,39 +1521,49 @@ struct _Variant_all_visit_results_same<_Callable, _Meta_list<_IndexVectors...>,
...
 };
 
 template <>
 struct _Visit_strategy<0> {
-    template <class, class _Callable>
-    static constexpr _Variant_visit_result_t<_Callable> _Visit(
-        size_t, _Callable&& _Obj) { // dispatch a visitation with 4^0 potential states
+    template <class _Ret, class, class _Callable>
+    static constexpr _Ret _Visit(size_t, _Callable&& _Obj) { // dispatch a visitation with 4^0 potential states
         return static_cast<_Callable&&>(_Obj)();
```
Comment (CaseyCarter) [inline stl/inc/variant]: I think this needs something like:
```c++
if constexpr(is_void_v<_Ret>) {
    return static_cast<void>(static_cast<_Callable&&>(_Obj)());
} else {
    return static_cast<_Callable&&>(_Obj)();
}
```
so silly uses like `std::visit<void>([] { return 42; });` work.

File: stl/inc/variant (hunk 5 above)
Comment (CaseyCarter) [reply stl/inc/variant]: Yes it does, and yes we should. Let's make these `static_cast`s for now.

File: stl/inc/variant [hunk 11]
```diff
@@ -1601,6 +1626,51 @@ template <class _Ty>
...
+                          _As_variant<_Variants>...>::value,
+            "visit<R>() requires the result of all potential invocations to be implicitly convertible to R "
+            "(N4828 [variant.visit]/2).");
+    }
+
+    return _Visit_impl<_Size, _ListOfIndexVectors, _Ret>(
+        _STD forward<_Callable>(_Obj), _STD forward<_Variants>(_Args)...);
+}
+#else // _HAS_CXX20
 template <class _Callable, class... _Variants, class = void_t<_As_variant<_Variants>...>>
```
Comment (CaseyCarter) [inline stl/inc/variant]: IIUC this version of vanilla visit is identical to the version in the `_HAS_CXX20` block above, except that `_Visit_impl` is effectively inlined into this one. If that's the case, we could reduce duplication by deleting this version and moving the other version and `_Visit_impl` out of the `_HAS_CXX20` block so they're used in both language modes.

File: stl/inc/variant (hunk 3 above)
Comment (CaseyCarter) [reply stl/inc/variant]: I think we can use this implementation of `visit` in both C++17 and C++20 mode (see my comment on the `visit` overload outside the `_HAS_CXX20` block), in which case these should certainly stay `_Remove_cvref_t`s. I think I'd prefer to keep the two overloads consistent where possible so the true differences stand out, so I'd leave the other overload alone as well, but I could easily be convinced otherwise if there were a reason to prefer `remove_cvref_t` to `_Remove_cvref_t` other than simply "the former is the standard's name".

File: stl/inc/variant [hunk 12]
```diff
@@ -39,6 +39,12 @@ struct _All_same : true_type {};
 template <class _First, class... _Rest>
 struct _All_same<_First, _Rest...> : bool_constant<conjunction_v<is_same<_First, _Rest>...>> {}; // variadic is_same
 
+template <class...>
+struct _All_convertible : true_type {};
```
Comment (CaseyCarter) [inline stl/inc/variant]: I don't think `_All_convertible<>` with no target type makes sense as does `_All_same<>`. I suggest dropping the above and making the below the primary template.

File: stl/inc/variant (hunk 3 above)
Comment (NathanSWard) [reply stl/inc/variant]: I also agree that if we use the vanilla `std::visit()` in both C++17 and C++20 then we should use `_Remove_cvref_t` in `std::visit<R>()` to keep the consistency between not only the two `visit` variations, but also among the file.

File: stl/inc/variant (hunk 9 above)
[the remark this answers is not in the document]
Comment (NathanSWard) [reply stl/inc/variant]: If I instead pass the return type to `_Variant_dispatcher`'s template parameters instead of `_Variant_dispatcher::_Dispatch`, is this safe?

File: stl/inc/variant (hunk 8 above)
Comment (NathanSWard) [reply stl/inc/variant]: Should `_Dispatch2`  be wrapped in a `#if _CXX20` block while keeping `_Dispatch` with C++17?

File: stl/inc/variant [hunk 13]
```diff
@@ -1611,14 +1646,31 @@ constexpr _Variant_visit_result_t<_Callable, _As_variant<_Variants>...> visit(_C
...
         "visit() requires the result of all potential invocations to have the same type and value category "
-        "(N4828 [variant.visit]/2).");
+        "(N4835 [variant.visit]/2).");
 
-    constexpr int _Strategy =
-        _Size == 1 ? 0 : _Size <= 4 ? 1 : _Size <= 16 ? 2 : _Size <= 64 ? 3 : _Size <= 256 ? 4 : -1;
-    return _Visit_strategy<_Strategy>::template _Visit<_ListOfIndexVectors>(
-        _Variant_visit_index1(0, static_cast<_As_variant<_Variants>&>(_Args)...), static_cast<_Callable&&>(_Obj),
-        static_cast<_As_variant<_Variants>&&>(_Args)...);
+    return _Visit_impl<_Size, _ListOfIndexVectors, _Variant_visit_result_t<_Callable, _As_variant<_Variants>...>>(
```
Comment (StephanTLavavej) [inline stl/inc/variant]: Would this be clearer if you extracted `using _Ret = _Variant_visit_result_t<_Callable, _As_variant<_Variants>...>;`?

File: stl/inc/variant [hunk 14]
```diff
@@ -1460,29 +1465,35 @@ struct _Variant_dispatcher;
 
 template <size_t... _Is>
 struct _Variant_dispatcher<index_sequence<_Is...>> {
-    template <class _Callable, class... _Types, bool _Any_valueless = ((_Is == 0) || ...)>
-    _NODISCARD static constexpr _Variant_visit_result_t<_Callable, _Types...> _Dispatch(
-        _Callable&& _Obj, _Types&&... _Args) {
+    template <class _Ret, class _Callable, class... _Types, bool _Any_valueless = ((_Is == 0) || ...)>
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/variant]: The grammar for this case of *fold-expression* is `(` *cast-expression* *fold-operator* `...` `)`. *cast-expression* is *unary-expression* or `(` *type-id* `)` *cast-expression*. In other words, `(_Is == 0 || ...)` is not a cast expression.

I'm 99% certain you asked this same question in the PR that added `_Any_valueless`. ;)

File: stl/inc/variant [hunk 15]
```diff
@@ -1600,6 +1625,15 @@ template <class _Ty>
 using _As_variant = // Deduce variant specialization from a derived type
     decltype(_As_variant_(_STD declval<_Ty>()));
 
+template <size_t _Size, class _ListOfIndexVectors, class _Ret, class _Callable, class... _Variants>
```
Comment (BillyONeal) [inline stl/inc/variant]: Mild preference to be consistent about `_ListOfIndexVectors` and `_Ret` order. Here _ListOfIndexVectors is first, but _Visit2 has the opposite order.

Comment (CaseyCarter) [reply stl/inc/variant]: I've made `_Visit_impl` consistent with the other templates that pass `_Ret` before `_ListOfIndexVectors`. (Sorry if I'm stepping on your toes, @NathanSWard, I'm trying to polish everything up to get a clean test run.)

Comment (NathanSWard) [reply stl/inc/variant]: No worries!
Thanks for all of the help @CaseyCarter with my PR's 😄

**Comment by CaseyCarter** [comment]
All green to merge. Thanks, @NathanSWard - great work!

**Comment by BillyONeal** [comment]
Thanks!
